app/containers/LoginCallback/sagas.js

```diff
@@ -0,0 +1,27 @@
+/**
+ * 1. sets up auth library and waits for auth token
+ * 2. gets user profile
+ */
+
+import { call, put } from 'redux-saga/effects';
+import { browserHistory } from 'react-router';
+import { loginCallback, getProfile } from './lib';
+import { loginSuccess, loginError } from './actions';
+
+function* loginCallbackSaga() {
+  // TODO take the proper event so that this saga is run exactly once (when loading route)
```

Comment:
````
export function* loginSaga() {

}

// You can escape using this watcher
export function* loginWatcher() {
  let action;
  while (action = yield take(SOME_ACTION_TYPE)) {
    yield call(loginSaga, action);
  }
}

export function* loginObserver() {
  const loginWatcher = yield fork(loginWatcher);
  const action = yield take(LOCATION_CHANGE);
  // Check condition of action if it's needed
  yield cancel(loginWatcher); // cancel task instance
}
````
I don't know if it helps but this is a pattern I did once when I encountered a similar problem. You can escape having loginWatcher and have 2 generators. It might seem bit hacky I guess.